language/libra-vm/src/logger.rs

```diff
@@ -0,0 +1,126 @@
+    block_id: Option<HashValue>,
+    // StateViewId::ChunkExecution - state sync
+    first_version: Option<Version>,
+    // StateViewId::TransactionValidation - validation
+    base_version: Option<Version>,
+
+    // transaction position in the list of transactions in the block,
+    // 0 if the transaction is not part of a block (i.e. validation).
+    txn_id: usize,
+}
+
+impl LogSchema {
+    pub fn new(view_id: StateViewId, txn_id: usize) -> Self {
+        match view_id {
+            StateViewId::BlockExecution { block_id } => Self {
+                name: LogEntry::Execution,
+                block_id: Some(block_id),
+                first_version: None,
+                base_version: None,
+                txn_id,
+            },
+            StateViewId::ChunkExecution { first_version } => Self {
+                name: LogEntry::Execution,
+                block_id: None,
+                first_version: Some(first_version),
+                base_version: None,
+                txn_id,
+            },
+            StateViewId::TransactionValidation { base_version } => Self {
+                name: LogEntry::Validation,
+                block_id: None,
+                first_version: None,
+                base_version: Some(base_version),
+                txn_id,
+            },
+            StateViewId::Miscellaneous => Self {
+                name: LogEntry::Miscellaneous,
+                block_id: None,
+                first_version: None,
+                base_version: None,
+                txn_id,
+            },
+        }
+    }
+}
+
+#[derive(Clone, Copy, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum LogEntry {
+    Execution,
+    Validation,
+    Miscellaneous, // usually testing
+}
+
+#[derive(Clone)]
+pub struct LibraLogger {
+    schema: LogSchema,
+}
+
+impl LibraLogger {
+    pub(crate) fn new(view_id: StateViewId, txn_id: usize) -> Self {
+        Self {
+            schema: LogSchema::new(view_id, txn_id),
+        }
+    }
+}
+
+impl Logger for LibraLogger {
+    // A critical log for the adapter, increment the `CRITICAL_ERRORS` monitor event so we can
+    // hook an alert when a critical occurs
+    fn crit(&self, message: &str) {
+        CRITICAL_ERRORS.inc();
+        error!(self.schema.clone(), "{}", message);
+    }
+
+    // REVIEW: consider hooking up a monitor event that we can use with a threshold
+    // over a period of time to raise an alert
+    fn error(&self, message: &str) {
+        error!(self.schema.clone(), "{}", message);
+    }
+
+    fn warn(&self, message: &str) {
+        warn!(self.schema.clone(), "{}", message);
+    }
+
+    fn info(&self, message: &str) {
+        info!(self.schema.clone(), "{}", message);
+    }
+
+    fn debug(&self, message: &str) {
+        debug!(self.schema.clone(), "{}", message);
+    }
+}
+
+// Helper `Logger` implementation that does nothing
+#[derive(Clone)]
+pub struct NoLogLogger;
```

Comment:
Should this be in `language/move-vm/types/src/logger.rs`?

Reply:
good deal, done